File: src/main/java/no/novari/flyt/integration/IntegrationController.java
package no.novari.flyt.integration;

import jakarta.validation.ConstraintViolation;
import lombok.extern.slf4j.Slf4j;
import no.novari.flyt.integration.model.dtos.IntegrationDto;
import no.novari.flyt.integration.model.dtos.IntegrationPatchDto;
import no.novari.flyt.integration.model.dtos.IntegrationPostDto;
import no.novari.flyt.integration.validation.IntegrationValidatorFactory;
import no.novari.flyt.integration.validation.ValidationErrorsFormattingService;
import no.novari.flyt.resourceserver.security.user.UserAuthorizationService;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import java.util.Collection;
import java.util.Set;

import static no.novari.flyt.resourceserver.UrlPaths.INTERNAL_API;


@RestController
@RequestMapping(INTERNAL_API + "/integrasjoner")
@Slf4j
public class IntegrationController {

    private final IntegrationService integrationService;
    private final IntegrationValidatorFactory integrationValidatorFactory;
    private final ValidationErrorsFormattingService validationErrorsFormattingService;
    private final UserAuthorizationService userAuthorizationService;

    public IntegrationController(
            IntegrationService integrationService,
            IntegrationValidatorFactory integrationValidatorFactory,
            ValidationErrorsFormattingService validationErrorsFormattingService,
            UserAuthorizationService userAuthorizationService) {
        this.integrationService = integrationService;
        this.integrationValidatorFactory = integrationValidatorFactory;
        this.validationErrorsFormattingService = validationErrorsFormattingService;
        this.userAuthorizationService = userAuthorizationService;
    }

    @GetMapping
    public ResponseEntity<Collection<IntegrationDto>> getIntegrations(
            @AuthenticationPrincipal Authentication authentication,
            @RequestParam(required = false) Long sourceApplicationId
    ) {
        return getResponseEntityIntegrations(authentication, sourceApplicationId);
    }

    @GetMapping(params = {"side", "antall", "sorteringFelt", "sorteringRetning"})
    public ResponseEntity<Page<IntegrationDto>> getIntegrations(
            @AuthenticationPrincipal Authentication authentication,
            @RequestParam(name = "side") int page,
            @RequestParam(name = "antall") int size,
            @RequestParam(name = "sorteringFelt") String sortProperty,
            @RequestParam(name = "sorteringRetning") Sort.Direction sortDirection,
            @RequestParam(required = false) Long sourceApplicationId
    ) {
        PageRequest pageRequest = PageRequest
                .of(page, size)
                .withSort(sortDirection, sortProperty);

        return getResponseEntityIntegrations(authentication, pageRequest, sourceApplicationId);
    }

    private ResponseEntity<Collection<IntegrationDto>> getResponseEntityIntegrations(
            Authentication authentication,
            Long sourceApplicationId
    ) {
        Set<Long> sourceApplicationIds =
                userAuthorizationService.getUserAuthorizedSourceApplicationIds(authentication);

        if (sourceApplicationId != null) {
            if (!sourceApplicationIds.contains(sourceApplicationId)) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
            }
            return ResponseEntity.ok(integrationService.findAllBySourceApplicationIds(Set.of(sourceApplicationId)));
        }

        return ResponseEntity.ok(integrationService.findAllBySourceApplicationIds(sourceApplicationIds));

    }

    private ResponseEntity<Page<IntegrationDto>> getResponseEntityIntegrations(
            Authentication authentication,
            Pageable pageable,
            Long sourceApplicationId
    ) {
        Set<Long> sourceApplicationIds =
                userAuthorizationService.getUserAuthorizedSourceApplicationIds(authentication);

        if (sourceApplicationId != null) {
            if (!sourceApplicationIds.contains(sourceApplicationId)) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
            }
            Page<IntegrationDto> allBySourceApplicationId = integrationService.findAllBySourceApplicationIds(
                    Set.of(sourceApplicationId), pageable
            );
            return ResponseEntity.ok(allBySourceApplicationId);
        }

        Page<IntegrationDto> allBySourceApplicationIds = integrationService.findAllBySourceApplicationIds(
                sourceApplicationIds, pageable
        );
        return ResponseEntity.ok(allBySourceApplicationIds);
    }

    @GetMapping("{integrationId}")
    public ResponseEntity<IntegrationDto> getIntegration(
            @AuthenticationPrincipal Authentication authentication,
            @PathVariable Long integrationId
    ) {
        IntegrationDto integrationDto = integrationService.findById(integrationId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));

        userAuthorizationService.checkIfUserHasAccessToSourceApplication(
                authentication,
                integrationDto.getSourceApplicationId());

        return ResponseEntity.ok(integrationDto);
    }

    @PostMapping
    public ResponseEntity<IntegrationDto> postIntegration(
            @AuthenticationPrincipal Authentication authentication,
            @RequestBody IntegrationPostDto integrationPostDto
    ) {
        userAuthorizationService.checkIfUserHasAccessToSourceApplication(
                authentication,
                integrationPostDto.getSourceApplicationId());

        validatePost(integrationPostDto);
        return ResponseEntity.ok(integrationService.save(integrationPostDto));
    }

    private void validatePost(IntegrationPostDto integrationPostDto) {
        Set<ConstraintViolation<IntegrationPostDto>> constraintViolations = integrationValidatorFactory
                .getValidator()
                .validate(integrationPostDto);
        if (!constraintViolations.isEmpty()) {
            throw new ResponseStatusException(
                    HttpStatus.UNPROCESSABLE_ENTITY,
                    validationErrorsFormattingService.format(constraintViolations)
            );
        }

        if (integrationService.existsIntegrationBySourceApplicationIdAndSourceApplicationIntegrationId(
                integrationPostDto.getSourceApplicationId(),
                integrationPostDto.getSourceApplicationIntegrationId()
        )) {
            throw new ResponseStatusException(HttpStatus.CONFLICT);
        }
    }

    @PatchMapping("{integrationId}")
    public ResponseEntity<IntegrationDto> patchIntegration(
            @AuthenticationPrincipal Authentication authentication,
            @PathVariable Long integrationId,
            @RequestBody IntegrationPatchDto integrationPatchDto
    ) {
        IntegrationDto existingIntegration = integrationService.findById(integrationId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));

        userAuthorizationService.checkIfUserHasAccessToSourceApplication(authentication, existingIntegration.getSourceApplicationId());

        IntegrationDto.IntegrationDtoBuilder integrationDtoBuilder = existingIntegration.toBuilder();

        integrationPatchDto.getDestination().ifPresent(integrationDtoBuilder::destination);
        integrationPatchDto.getState().ifPresent(integrationDtoBuilder::state);
        integrationPatchDto.getActiveConfigurationId().ifPresent(integrationDtoBuilder::activeConfigurationId);

        validatePatchResult(integrationId, integrationDtoBuilder.build());

        return ResponseEntity.ok(integrationService.updateById(integrationId, integrationPatchDto));
    }

    private void validatePatchResult(Long integrationId, IntegrationDto integrationDto) {
        Set<ConstraintViolation<IntegrationDto>> constraintViolations = integrationValidatorFactory
                .getPatchValidator(
                        integrationId,
                        integrationDto.getActiveConfigurationId()
                )
                .validate(integrationDto);
        if (!constraintViolations.isEmpty()) {
            throw new ResponseStatusException(
                    HttpStatus.UNPROCESSABLE_ENTITY,
                    validationErrorsFormattingService.format(constraintViolations)
            );
        }
    }

}

File: src/main/kotlin/no/novari/flyt/integration/api/IntegrationController.kt
package no.novari.flyt.integration.api

import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.Parameter
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.responses.ApiResponses
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.Valid
import no.novari.flyt.integration.api.dto.IntegrationDto
import no.novari.flyt.integration.api.dto.IntegrationPageResponse
import no.novari.flyt.integration.api.dto.IntegrationPatchDto
import no.novari.flyt.integration.api.dto.IntegrationPostDto
import no.novari.flyt.integration.application.IntegrationService
import no.novari.flyt.integration.application.IntegrationUpdateValidationService
import no.novari.flyt.integration.web.ConflictException
import no.novari.flyt.integration.web.ForbiddenWithoutBodyException
import no.novari.flyt.integration.web.NotFoundException
import no.novari.flyt.webresourceserver.UrlPaths.INTERNAL_API
import no.novari.flyt.webresourceserver.security.user.UserAuthorizationService
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.security.core.Authentication
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("$INTERNAL_API/integrasjoner")
@Tag(name = "Integrations", description = "Management of Flyt integration definitions.")
class IntegrationController(
    private val integrationService: IntegrationService,
    private val integrationUpdateValidationService: IntegrationUpdateValidationService,
    private val userAuthorizationService: UserAuthorizationService,
) {
    @GetMapping
    @Operation(summary = "List integrations", operationId = "listIntegrations")
    fun listIntegrations(
        authentication: Authentication,
        @Parameter(description = "Filter by source application identifier")
        @RequestParam(required = false) sourceApplicationId: Long?,
    ): Collection<IntegrationDto> {
        val sourceApplicationIds = resolveAuthorizedSourceApplicationIds(authentication, sourceApplicationId)
        return integrationService.findAllBySourceApplicationIds(sourceApplicationIds)
    }

    @GetMapping(params = ["side", "antall", "sorteringFelt", "sorteringRetning"])
    @Operation(summary = "List integrations with pagination", operationId = "listIntegrationsPaginated")
    fun listIntegrationsPaginated(
        authentication: Authentication,
        @Parameter(description = "Zero-based page number")
        @RequestParam(name = "side") page: Int,
        @Parameter(description = "Number of integrations per page")
        @RequestParam(name = "antall") size: Int,
        @Parameter(description = "Property to sort by")
        @RequestParam(name = "sorteringFelt") sortProperty: String,
        @Parameter(description = "Sort direction")
        @RequestParam(name = "sorteringRetning") sortDirection: Sort.Direction,
        @Parameter(description = "Filter by source application identifier")
        @RequestParam(required = false) sourceApplicationId: Long?,
    ): IntegrationPageResponse {
        val sourceApplicationIds = resolveAuthorizedSourceApplicationIds(authentication, sourceApplicationId)
        val pageRequest = PageRequest.of(page, size).withSort(sortDirection, sortProperty)
        val integrations = integrationService.findAllBySourceApplicationIds(sourceApplicationIds, pageRequest)
        return IntegrationPageResponse(
            content = integrations.content,
            totalElements = integrations.totalElements,
            totalPages = integrations.totalPages,
        )
    }

    @GetMapping("{integrationId}")
    @Operation(summary = "Get an integration")
    @ApiResponses(
        value = [
            ApiResponse(responseCode = "200", description = "Integration found"),
            ApiResponse(responseCode = "403", description = "Source application access denied"),
            ApiResponse(responseCode = "404", description = "Integration not found"),
        ],
    )
    fun getIntegrationById(
        authentication: Authentication,
        @Parameter(description = "Integration identifier")
        @PathVariable integrationId: Long,
    ): IntegrationDto {
        val integration =
            integrationService.findById(integrationId)
                ?: throw NotFoundException()

        userAuthorizationService.checkIfUserHasAccessToSourceApplication(
            authentication,
            integration.sourceApplicationId,
        )

        return integration
    }

    @PostMapping
    @Operation(summary = "Create an integration")
    @ApiResponses(
        value = [
            ApiResponse(responseCode = "200", description = "Integration created"),
            ApiResponse(responseCode = "403", description = "Source application access denied"),
            ApiResponse(responseCode = "409", description = "Integration already exists"),
            ApiResponse(responseCode = "422", description = "Invalid integration"),
        ],
    )
    fun createIntegration(
        authentication: Authentication,
        @Valid @RequestBody integrationPostDto: IntegrationPostDto,
    ): IntegrationDto {
        ensureIntegrationDoesNotAlreadyExist(integrationPostDto)

        userAuthorizationService.checkIfUserHasAccessToSourceApplication(
            authentication,
            requireNotNull(integrationPostDto.sourceApplicationId),
        )

        return integrationService.save(integrationPostDto)
    }

    @PatchMapping("{integrationId}")
    @Operation(summary = "Update an integration")
    @ApiResponses(
        value = [
            ApiResponse(responseCode = "200", description = "Integration updated"),
            ApiResponse(responseCode = "403", description = "Source application access denied"),
            ApiResponse(responseCode = "404", description = "Integration not found"),
            ApiResponse(responseCode = "422", description = "Invalid integration"),
        ],
    )
    fun updateIntegration(
        authentication: Authentication,
        @Parameter(description = "Integration identifier")
        @PathVariable integrationId: Long,
        @RequestBody integrationPatchDto: IntegrationPatchDto,
    ): IntegrationDto {
        val existingIntegration =
            integrationService.findById(integrationId)
                ?: throw NotFoundException()

        userAuthorizationService.checkIfUserHasAccessToSourceApplication(
            authentication,
            existingIntegration.sourceApplicationId,
        )

        val updatedIntegration =
            existingIntegration.copy(
                destination = integrationPatchDto.destination ?: existingIntegration.destination,
                state = integrationPatchDto.state ?: existingIntegration.state,
                activeConfigurationId =
                    integrationPatchDto.activeConfigurationId ?: existingIntegration.activeConfigurationId,
            )

        integrationUpdateValidationService.validate(integrationId, updatedIntegration)

        return integrationService.updateById(integrationId, integrationPatchDto)
    }

    private fun resolveAuthorizedSourceApplicationIds(
        authentication: Authentication,
        sourceApplicationId: Long?,
    ): Set<Long> {
        val candidateSourceApplicationIds =
            sourceApplicationId?.let(::setOf) ?: integrationService.findDistinctSourceApplicationIds()
        val authorizedSourceApplicationIds =
            userAuthorizationService.getUserAuthorizedSourceApplicationIds(
                authentication,
                candidateSourceApplicationIds,
            )

        if (sourceApplicationId != null && !authorizedSourceApplicationIds.contains(sourceApplicationId)) {
            throw ForbiddenWithoutBodyException()
        }

        return authorizedSourceApplicationIds
    }

    private fun ensureIntegrationDoesNotAlreadyExist(integrationPostDto: IntegrationPostDto) {
        if (
            integrationService.existsIntegrationBySourceApplicationIdAndSourceApplicationIntegrationId(
                requireNotNull(integrationPostDto.sourceApplicationId),
                requireNotNull(integrationPostDto.sourceApplicationIntegrationId),
            )
        ) {
            throw ConflictException()
        }
    }
}
